[TACHYON-645] Use createNewClassInstance to improve EvictorFactory

lindong28 opened the pull request:
https://tachyon.atlassian.net/browse/TACHYON-645

AmplabJenkins commented:
Can one of the admins verify this patch?

uronce-cc commented on servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java:
```diff
@@ -16,11 +16,14 @@
 import java.util.Collections;
 import java.util.List;
 
+
```
empty line

uronce-cc commented on servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java:
```diff
@@ -87,7 +92,10 @@ public final void before() throws Exception {
+    Map<String, String> props = new LinkedHashMap<String, String>();
+    props.put(Constants.WORKER_EVICT_STRATEGY_TYPE, mEvictorClassName);
+    mEvictor = EvictorFactory.create(mManagerView, new TachyonConf(props));
+
```
empty line.

uronce-cc commented on servers/src/main/java/tachyon/EvictorConstants.java:
```diff
@@ -13,24 +13,19 @@
-   * Evict blocks continually until a StorageDir has enough space
-   */
-  GREEDY,
+public class EvictorConstants {
```
It seems fine to keep its file path and keep in `tachyon.worker.block.evictor`?

uronce-cc commented on servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java:
```diff
@@ -66,16 +73,14 @@
-    }
+    list.add(new Object[] {EvictorConstants.GREEDY_EVICTOR_CLASS.getName()});
+    list.add(new Object[] {EvictorConstants.LRU_EVICTOR_CLASS.getName()});
+
```
Each new evictor needs to add its own class name here for evictor contract test now, before change, they only need to be defined in the `EvictorType`. 

What about having a `public static final Class<? extends Evictor>[] evictorTypes = new xxx { GREEDY_EVICTOR_CLASS, LRU_EVICTOR_CLASS }; ` in `EvictorConstants`, so that when a new evictor is added, we add its type to `EvictorConstants` and `evictorTypes`, and in `data()`, we can use `evictorTypes` ?

uronce-cc commented:
@lindong28 Could you also update the default configuration values here: https://github.com/amplab/tachyon/blob/master/common/src/main/resources/tachyon-default.properties#L73, and the documentation here: https://github.com/amplab/tachyon/blob/master/docs/Configuration-Settings.md? Thanks!

apc999 commented:
@gpang  @uronce-cc  please review this PR.

lindong28 commented:
@uronce-cc Thanks for taking time to review the patch.

I have addressed most comments in the 2nd commit. But I am not sure if I understand your comment on configuration, since I didn't add new configuration. tachyon.worker.evict.strategy is already documented and it has default value LRU. Did I miss something here?

lindong28 commented:
Oh.. My bad. I just realized your point on configuration. Will change it.

gpang commented:
ok to test

gpang commented on common/src/main/resources/tachyon-default.properties:
```diff
@@ -70,7 +70,7 @@ tachyon.worker.memory.size=128MB
 tachyon.worker.network.netty.watermark.low=8KB
 tachyon.worker.allocate.strategy=MAX_FREE
-tachyon.worker.evict.strategy=LRU
+tachyon.worker.evict.strategy=tachyon.worker.block.evictor.LRUEvictor
```
Could you rename this property as "tachyon.worker.evict.strategy.class"? (basically, add the suffix ".class")

Could you also rename the Constant to match the property name?

lindong28 replied:
Fixed. It is OK to have this interface change, right?

gpang commented on servers/src/main/java/tachyon/worker/block/evictor/EvictorFactory.java:
```diff
@@ -15,33 +15,31 @@
+  public static Evictor create(final BlockMetadataManagerView view, TachyonConf conf) {
+    try {
+      return CommonUtils.createNewClassInstance(
+          conf.getClass(Constants.WORKER_EVICT_STRATEGY_TYPE, EvictorConstants.LRU_EVICTOR_CLASS),
```
I think here, you can just hardcode the default "LRUEvictor.class", and you can get rid of the EvictorConstants class. That is only used in the tests, and in the tests you can do something similar to PR #1133 to get all the classes implementing the Evictor interface.

lindong28 replied:
Ah, I thought of enumerating the subclass of Evictor interface, but thought this is an overkill.

I used it in the latest patch using almost the same code in #1133. Maybe we should move this code into a utility function to avoid code duplication?

gpang commented on servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java:
```diff
@@ -40,6 +40,8 @@
 
   public GreedyEvictor() {}
```
You can just get rid of this constructor.

gpang commented on servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java:
```diff
@@ -40,6 +40,8 @@
 
   public GreedyEvictor() {}
 
+  public GreedyEvictor(BlockMetadataManagerView view) {}
```
Please leave a simple comment saying this evictor strategy doesn't need the view.

gpang commented on servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java:
```diff
@@ -56,7 +62,7 @@
   private BlockMetadataManager mMetaManager;
   private BlockMetadataManagerView mManagerView;
-  private EvictorType mEvictorType;
+  private String mEvictorClassName;
```
I think this can be final.

gpang commented on servers/src/test/java/tachyon/worker/block/evictor/EvictorTest.java:
```diff
@@ -87,7 +91,9 @@ public final void before() throws Exception {
     mTestDir = tiers.get(TEST_TIER_LEVEL).getDir(TEST_DIR);
-    mEvictor = EvictorFactory.create(mEvictorType, mManagerView);
+    Map<String, String> props = new LinkedHashMap<String, String>();
+    props.put(Constants.WORKER_EVICT_STRATEGY_TYPE, mEvictorClassName);
```
You can create a TachyonConf(), and then set() the evictor class name, instead of filling the Map directly.

gpang commented:
Thanks for the PR! I left some comments. Also, could you take a look at PR #1133 and try to match the similar style? That PR is the analogous change to the Allocator, so it would be good to have similar changes to the evictor.

Also, could you resolve the merge conflicts? Thanks!

lindong28 commented:
@gpang Thanks for taking time to review the patch! I have addressed most comments. Please see my reply under your comments.

lindong28 commented:
@gpang Can you have a look at the updated patch? I haven't resolved the merge conflict yet. I worry that either git merge or git rebase will make it hard for reviewer to review changes, no? How about I resolve the conflicts after the patch is approved?

gpang commented on servers/src/main/java/tachyon/EvictorConstants.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Enumerate evictors
+ */
+public class EvictorConstants {
```
I don't think this file is necessary anymore.

gpang commented on servers/src/main/java/tachyon/worker/block/evictor/EvictorFactory.java:
```diff
@@ -15,33 +15,30 @@
-        return new LRUEvictor(managerView);
-      default:
-        return new LRUEvictor(managerView);
+  public static Evictor create(final BlockMetadataManagerView view, TachyonConf conf) {
```
Just to stay consistent with the Allocator factory in PR #1133 , maybe this should be moved to an inner class of Evictor, and called createEvictor().

To be consistent, please re-order the parameters to be (conf, view)

gpang commented:
I left a few more comments. I think you can resolve merge conflicts now, and it should be fine. Thanks!

lindong28 commented:
@gpang  I have rebased my patch onto latest version of tachyon and squashed the patch into one commit. Please have a look. Thanks!

gpang commented:
Please fix the compilation error.

Also, next time it is not necessary to squash all the commits into one, since the history is now gone.

calvinjia commented:
Jenkins, test this please.

lindong28 commented:
@gpang Oh, my bad. I will not squash commits next time.

Compilation succeeds on my machine. Jenkins doesn't show what's going wrong with compilation either. Can you help me with it?

gpang commented:
Ok. Did you try "mvn clean", before "mvn install"?

Here is what jenkins says:

[ERROR] /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/src/test/java/tachyon/worker/block/evictor/PartialLRUEvictorTest.java:[54,38] cannot find symbol
  symbol:   variable EvictorType
  location: class tachyon.worker.block.evictor.PartialLRUEvictorTest
[ERROR] /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/src/test/java/tachyon/worker/block/evictor/PartialLRUEvictorTest.java:[54,16] cannot find symbol
  symbol:   variable EvictorFactory
  location: class tachyon.worker.block.evictor.PartialLRUEvictorTest
[INFO] 2 errors

lindong28 commented:
I executed "mvn clean && mvn compile && mvn test" before submitting the
code.

It turns out that "mvn install" and "mvn test" will report the error but
"mvn compile" succeeds. I think this is problematic, and we should fix "mvn
compile" to also compile the testcase and report error if needed. What do
you think?

On Wed, Jul 22, 2015 at 1:18 PM, Gene Pang <notifications@github.com> wrote:

> Ok. Did you try "mvn clean", before "mvn install"?
>
> Here is what jenkins says:
>
> [ERROR]
> /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/src/test/java/tachyon/worker/block/evictor/PartialLRUEvictorTest.java:[54,38]
> cannot find symbol
> symbol: variable EvictorType
> location: class tachyon.worker.block.evictor.PartialLRUEvictorTest
> [ERROR]
> /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/src/test/java/tachyon/worker/block/evictor/PartialLRUEvictorTest.java:[54,16]
> cannot find symbol
> symbol: variable EvictorFactory
> location: class tachyon.worker.block.evictor.PartialLRUEvictorTest
> [INFO] 2 errors
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1144#issuecomment-123850216>.
>

lindong28 commented:
I executed "mvn clean && mvn compile && mvn test" before submitting the code.

It turns out that "mvn install" and "mvn test" will report the error but "mvn compile" succeeds. I think this is problematic, and we should fix "mvn compile" to also compile the testcase and report error if needed. What do you think?

gpang commented:
LGTM

lindong28 commented:
@g

apc999 commented:
+1 LGTM
